src/promptflow/promptflow/executor/_line_execution_process_pool.py

```diff
@@ -152,121 +91,199 @@ def __init__(
+            self._monitor_pool.close()
+            self._monitor_pool.join()
+
+    def _get_process_info(self, index):
+        while True:
+            try:
+                process_id = self._process_info[index].process_id
+                process_name = self._process_info[index].process_name
+                return (index, process_id, process_name)
+            except KeyError:
+                continue
+            except Exception as e:
+                bulk_logger.warning(f"Unexpected error occurred while get process info. Exception: {e}")
+
+    def handle_line_timeout(self, line_number, timeout_time, inputs, run_id, start_time, result_list):
+        bulk_logger.warning(f"Line {line_number} timeout after {timeout_time} seconds.")
+        ex = LineExecutionTimeoutError(line_number, timeout_time)
+        result = self._generate_line_result_for_exception(inputs, run_id, line_number, self._flow_id, start_time, ex)
+        result_list.append(result)
 
-    def _timeout_process_wrapper(self, run_start_time: datetime, task_queue: Queue, timeout_time, result_list):
-        healthy_ensured_process = HealthyEnsuredProcess(self._executor_creation_func, self.context)
-        healthy_ensured_process.start_new(task_queue)
+    def handle_process_crashed(self, line_number, inputs, run_id, start_time, result_list):
+        bulk_logger.warning(f"Process crashed while executing line {line_number},")
+        ex = ProcessCrashError(line_number)
+        result = self._generate_line_result_for_exception(inputs, run_id, line_number, self._flow_id, start_time, ex)
+        result_list.append(result)
 
-        if not healthy_ensured_process.process.is_alive():
-            return
+    def _monitor_process_alive(self, process_id):
+        return not psutil.pid_exists(process_id)
+
+    def _handle_output_queue_messages(self, output_queue, result_list):
+        try:
+            message = output_queue.get(timeout=1)
+            if isinstance(message, LineResult):
+                message = self._process_multimedia(message)
+                result_list.append(message)
+                return True
+            elif isinstance(message, FlowRunInfo):
+                self._storage.persist_flow_run(message)
+            elif isinstance(message, NodeRunInfo):
+                self._storage.persist_node_run(message)
+        except queue.Empty:
+            pass
+        return False
+
+    def _monitor_workers_and_process_tasks_in_thread(
+        self, task_queue: Queue, timeout_time, result_list, index, input_queue, output_queue
+    ):
+        index, process_id, process_name = self._get_process_info(index)
 
         while True:
             try:
+                # Get task from task_queue
                 args = task_queue.get(timeout=1)
             except queue.Empty:
-                healthy_ensured_process.end()
+                self._processes_manager.end_process(index)
+                while psutil.pid_exists(process_id):
```

Comment:
Why this? Seems to be an infinite loop if process_id and index doesn't match.

Reply:
In fork mode, the main process and the sub spawn process communicate through _process_info. We need to ensure the process has been killed before returning. Otherwise, it may cause the main process have exited but the spawn process is still alive. At this time, a connection error will be reported.

process_id and index are match, process_info uses index as key, process_id and process_name as value.
![image](https://github.com/microsoft/promptflow/assets/115983968/10a55499-3f9d-41dc-ba6a-e962205bba41)